File: Editorial/52_N_Queens_II/JavaScript/1_Approach_Backtracking.js
var totalNQueens = function (n) {
    function backtrack(row, diagonals, anti_diagonals, cols) {
        if (row == n) {
            return 1;
        }
        let solutions = 0;
        for (let col = 0; col < n; col++) {
            let curr_diagonal = row - col;
            let curr_anti_diagonal = row + col;
            // If the queen is not placeable
            if (
                cols.has(col) ||
                diagonals.has(curr_diagonal) ||
                anti_diagonals.has(curr_anti_diagonal)
            ) {
                continue;
            }
            cols.add(col);
            diagonals.add(curr_diagonal);
            anti_diagonals.add(curr_anti_diagonal);
            solutions += backtrack(row + 1, diagonals, anti_diagonals, cols);
            cols.delete(col);
            diagonals.delete(curr_diagonal);
            anti_diagonals.delete(curr_anti_diagonal);
        }
        return solutions;
    }
    return backtrack(0, new Set(), new Set(), new Set());
};

File: Editorial/0000-0100/0052_N_Queens_II/Go/1_Approach_Backtracking.go
func totalNQueens(n int) int {
    size := n
    var backtrack func(int, map[int]bool, map[int]bool, map[int]bool) int
    backtrack = func(row int, diagonals, antiDiagonal, cols map[int]bool) int {
        if row == size {
            return 1
        }
        solutions := 0
        for col := 0; col < size; col++ {
            currDiagonal := row - col
            currAntiDiagonal := row + col
            // If the queen is not placeable
            if cols[col] || diagonals[currDiagonal] ||
                antiDiagonal[currAntiDiagonal] {
                continue
            }
            cols[col] = true
            diagonals[currDiagonal] = true
            antiDiagonal[currAntiDiagonal] = true
            // Move on to the next row with the updated board state
            solutions += backtrack(row+1, diagonals, antiDiagonal, cols)
            delete(cols, col)
            delete(diagonals, currDiagonal)
            delete(antiDiagonal, currAntiDiagonal)
        }
        return solutions
    }
    return backtrack(
        0,
        make(map[int]bool),
        make(map[int]bool),
        make(map[int]bool),
    )
}
